pkg/storage/ingest/partition_offset_client.go

```diff
@@ -0,0 +1,194 @@
+			Buckets:                         prometheus.DefBuckets,
+		}, []string{"partition"}),
+	}
+}
+
+// FetchLastProducedOffset fetches and returns the last produced offset for a partition, or -1 if no record has
+// been ever produced in the partition. This function issues a single request, but the Kafka client used under the
+// hood may retry a failed request until the retry timeout is hit.
+func (p *partitionOffsetClient) FetchLastProducedOffset(ctx context.Context, partitionID int32) (_ int64, returnErr error) {
+	var (
+		startTime        = time.Now()
+		partitionIDLabel = strconv.Itoa(int(partitionID))
+
+		// Init metrics for a partition even if they're not tracked (e.g. failures).
+		requestsTotal   = p.lastProducedOffsetRequestsTotal.WithLabelValues(partitionIDLabel)
+		requestsLatency = p.lastProducedOffsetLatency.WithLabelValues(partitionIDLabel)
+		failuresTotal   = p.lastProducedOffsetFailuresTotal.WithLabelValues(partitionIDLabel)
+	)
+
+	requestsTotal.Inc()
+	defer func() {
+		// We track the latency also in case of error, so that if the request times out it gets
+		// pretty clear looking at latency too.
+		requestsLatency.Observe(time.Since(startTime).Seconds())
+
+		if returnErr != nil {
+			failuresTotal.Inc()
+		}
+	}()
+
+	offset, err := p.fetchPartitionOffset(ctx, partitionID, kafkaOffsetEnd)
+	if err != nil {
+		return 0, err
+	}
+
+	// The offset we get is the offset at which the next message will be written, so to get the last produced offset
+	// we have to subtract 1. See DESIGN.md for more details.
+	return offset - 1, nil
+}
+
+// FetchPartitionStartOffset fetches and returns the start offset for a partition. This function returns 0 if no record has
+// been ever produced in the partition. This function issues a single request, but the Kafka client used under the
+// hood may retry a failed request until the retry timeout is hit.
+func (p *partitionOffsetClient) FetchPartitionStartOffset(ctx context.Context, partitionID int32) (_ int64, returnErr error) {
+	var (
+		startTime        = time.Now()
+		partitionIDLabel = strconv.Itoa(int(partitionID))
+
+		// Init metrics for a partition even if they're not tracked (e.g. failures).
+		requestsTotal   = p.partitionStartOffsetRequestsTotal.WithLabelValues(partitionIDLabel)
+		requestsLatency = p.partitionStartOffsetLatency.WithLabelValues(partitionIDLabel)
+		failuresTotal   = p.partitionStartOffsetFailuresTotal.WithLabelValues(partitionIDLabel)
+	)
+
+	requestsTotal.Inc()
+	defer func() {
+		// We track the latency also in case of error, so that if the request times out it gets
+		// pretty clear looking at latency too.
+		requestsLatency.Observe(time.Since(startTime).Seconds())
+
+		if returnErr != nil {
+			failuresTotal.Inc()
+		}
+	}()
+
+	return p.fetchPartitionOffset(ctx, partitionID, kafkaOffsetStart)
+}
+
+func (p *partitionOffsetClient) fetchPartitionOffset(ctx context.Context, partitionID int32, position int64) (int64, error) {
```

Comment:
Could you add a comment on why didn't [`kadm.Client.ListStartOffsets`][1] and [`kadm.Client.ListEndOffsets`][2] work here, and we had to construct the low-level messages manually? We use the variant of those helper functions in the `Reader`, so maybe there is a room to simplify it here to.

[1]: https://pkg.go.dev/github.com/twmb/franz-go/pkg/kadm#Client.ListStartOffsets
[2]: https://pkg.go.dev/github.com/twmb/franz-go/pkg/kadm#Client.ListEndOffsets

Reply:
Yes I will. The reason is very simple: [kadm.Client.ListStartOffsets](https://pkg.go.dev/github.com/twmb/franz-go/pkg/kadm#Client.ListStartOffsets) and [kadm.Client.ListEndOffsets](https://pkg.go.dev/github.com/twmb/franz-go/pkg/kadm#Client.ListEndOffsets) fetch offsets for all partitions, not a single one. It's just more expensive to fetch for all partitions, but ingesters only need the offset of their own one.

Reply:
Will comment on it in the next PR.